### PR TITLE
fix: amend incorrect int bound check

### PR DESCRIPTION
**Pull request by gabrielmer**
int64 numbers range from `-9,223,372,036,854,775,808` to `9,223,372,036,854,775,807`.

When checking if an int64 is in a valid range, we have to check whether its absolute value is smaller or equal than `9,223,372,036,854,775,808` for negative numbers, and smaller or equal than `9,223,372,036,854,775,807` for positive numbers.

Without this fix, `9,223,372,036,854,775,808` is taken as a valid int64 number which is not.

### CODE REVIEW

**Review by Ivansete-status — APPROVED** [review]
LGTM! Thanks for it! :100:

**Comment by etan-status** [comment]
I think that a test for the edge case being addressed here would be great.

**Comment by gabrielmer** [comment]
> I think that a test for the edge case being addressed here would be great.

Great idea! Just added a couple tests and fixed some bugs I had missed :)

File: json_serialization/reader.nim [hunk 1]
```diff
@@ -178,6 +178,7 @@ func maxAbsValue(T: type[SomeInteger]): uint64 {.compileTime.} =
   elif T is int16: 32768'u64
   elif T is int32: 2147483648'u64
   elif T is int64: 9223372036854775808'u64
+  elif T is int: 9223372036854775808'u64
```
Comment (etan-status) [inline json_serialization/reader.nim:181]: that depends on `sizeof(int)`, i.e., i386 (32-bit cpu) uses `int32`.

wouldn't `high(T).uint64 + 1` work for `SomeSignedInt`? a bit weird that we have to hardcode these constants (even though this PR isn't the one introducing this methodology).

Comment (gabrielmer) [reply json_serialization/reader.nim:181]: Agree 100%, I'm also not sure why the values are hardcoded. I'll modify the function for something like what you suggested

**Review by etan-status — APPROVED** [review]

File: json_serialization/reader.nim [hunk 2]
```diff
@@ -587,28 +580,43 @@ proc readValue*[T](r: var JsonReader, value: var T)
...
     r.lexer.next()
 
-  elif value is SomeInteger:
+  elif value is SomeSignedInt:
     type TargetType = type(value)
-    const maxValidValue = maxAbsValue(TargetType)
+    let
+      isNegative = tok == tkNegativeInt
+      maxValidAbsValue =
+        if isNegative:
```
Comment (Ivansete-status) [inline json_serialization/reader.nim:588]: Are we sure this doesn't produce overflow?

Comment (gabrielmer) [reply json_serialization/reader.nim:588]: Great question! Yes, because of the casting to `uint64`. 

`TargetType.high` will always be strictly smaller than the highest value for a `uint64`, as `TargetType` has to be a signed integer.

So after casting it to `uint64`, adding 1 to it won't produce an overflow

Comment (gabrielmer) [reply json_serialization/reader.nim:588]: Sounds gooood :), will do now!

File: json_serialization/reader.nim [hunk 3]
```diff
@@ -187,7 +187,7 @@ proc readJsonNodeField(r: var JsonReader, field: var JsonNode)
   field = r.parseJsonNode()
 
 proc parseJsonNode(r: var JsonReader): JsonNode =
-  const maxIntValue = BiggestInt.high.uint64 + 1
+  const maxIntValue: uint64 = BiggestInt.high.uint64 + 1
```
Comment (Ivansete-status) [inline json_serialization/reader.nim:190]: `BiggestInt` means signed int?
A recurrent question: are we sure this doesn't overflow? :)

**Review by Ivansete-status — APPROVED** [review]
LGTM! Thanks!
Just added a persistent question xD

File: json_serialization/reader.nim (hunk 3 above)
Comment (gabrielmer) [reply json_serialization/reader.nim:190]: Ooh good one! `BiggestInt` is an alias for `int64` as per
https://github.com/status-im/Nim/blob/3c9b68dc157804885b14a1984efc25e8b7cc861d/lib/system/basic_types.nim#L36-L39

So yes, it means the biggest type of signed int :)
Do you think it's better to use int64 instead?

Comment (Ivansete-status) [reply json_serialization/reader.nim:190]: > Do you think it's better to use int64 instead?

Yes good point! I think is better to explicitly use `int64` to make sure it won't fail in the future if the `BiggestInt` definition changes to `int128` for instance, if that type ever exists in the future. 

Aside from that, will that work on 32 bit architecture? I'm thinking whether the `unit32` and `int32` types will suffice for us.

wdyt? @etan-status @gabrielmer

Comment (Ivansete-status) [reply json_serialization/reader.nim:190]: Please disregard this comment:
> Aside from that, will that work on 32 bit architecture? I'm thinking whether the unit32 and int32 types will suffice for us.

`int64` can work well in 32-bit architecture too.

Comment (gabrielmer) [reply json_serialization/reader.nim:190]: I have kind of mixed feelings.

On one hand, I like the naming of `BiggestInt` because of explaining that we intentionally use the biggest `int` possible and the hypothetical case of an int128 can be far fetched (although not negligible I guess).

On the other, yes, if at some point there gets to exist an `int128` it we'll have a bug. It would remind us to change the code to support the new type promptly, but it would fail at the beginning and that's never good.

@etan-status what's your take?

Comment (etan-status) [reply json_serialization/reader.nim:190]: can we use `BiggestUint` (or equivalent) instead of `uint64`? that would also solve the problem.

If it doesn't work, can put a `static: doAssert sizeof(uint64) >= sizeof(BiggestInt)` or similar, to have a compile-time check that the assumption holds. If it ever changes, there would be a compile-time error and you would get alerted to it.

File: json_serialization/reader.nim (hunk 2 above)
Comment (etan-status) [reply json_serialization/reader.nim:588]: The `TargetType.high` will always be strictly smaller is only true on 2-s complement, btw. which, all target architectures for Status are.

File: json_serialization/reader.nim (hunk 3 above)
Comment (gabrielmer) [reply json_serialization/reader.nim]: I like the solution! I switched the uses of `uint64` and `int64` to `BiggestUint` and `BiggestInt` respectively. Everything worked for me locally, hope it passes the CI.

Please let me know if this looks ok and if so then we should be good to go :))

Comment (Ivansete-status) [reply json_serialization/reader.nim]: > Please let me know if this looks ok and if so then we should be good to go :))

LGMT! Thank you guys! :clap:

**Review by etan-status — APPROVED** [review]
